File: Sites/Trump/client/src/types/operations.ts
// Phase 03 — staff operations types (shifts, ownership, notifications, owner ops, audit).

export interface ShiftRow {
  id: number;
  username: string;
  role: string;
  status: 'active' | 'ended';
  startedAt: string;
  endedAt: string | null;
  startedBy: string;
  endedBy: string;
  endReason: string;
  assignedTables?: unknown;
  ordersHandled: number;
  revenueHandled: number;
  // Phase 2 (Waiter Experience) — tipsHandled lives inside responseMetrics
  // (a Json column) rather than as its own DB column, so it survives both the
  // live view and the ended-shift summary without a schema change.
  responseMetrics?: { tasksResolved?: number; tipsHandled?: number } | null;
}

export interface ShiftStatus {
  active: boolean;
  shift?: ShiftRow;
  ordersHandled?: number;
  revenueHandled?: number;
  tipsHandled?: number;
  responseMetrics?: { tasksResolved?: number; tipsHandled?: number };
}

export interface OwnershipRow {
  id: number;
  tableId: string;
  waiterName: string;
  status: string;
  assignedAt: string;
  releasedAt: string | null;
  changeType: string;
  assignedBy: string;
  previousWaiter: string;
  reason: string;
}

export interface NotificationRow {
  id: number;
  source: string;
  title: string;
  body: string;
  priority: number;
  recipientRole: string;
  recipientUser: string;
  tableId: string;
  readAt: string | null;
  createdAt: string;
}

export interface WaiterPerf {
  waiter: string;
  orders: number;
  revenue: number;
  tables: number;
}

export interface OpsSnapshot {
  generatedAt: string;
  activeWaiters: { username: string; startedAt: string }[];
  activeManagers: { username: string; startedAt: string }[];
  activeWaiterCount: number;
  activeManagerCount: number;
  openTables: number;
  tablesOwned: number;
  ordersToday: number;
  revenueToday: number;
  reservationsToday: number;
  waiterPerformance: WaiterPerf[];
  notifications: { unread: number };
  systemHealth: { ok: boolean };
}

export interface AuditRow {
  id: number;
  actor: string;
  actorRole: string;
  action: string;
  targetType: string;
  targetId: string;
  summary: string;
  reason: string;
  metadata?: unknown;
  createdAt: string;
}

File: luxury/backend/app/models/operations.py
"""Operations models — Shift, AuditLog, Notification, WaiterTask.

Map to existing Prisma tables for operational tracking.
"""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import DateTime, Float, Integer, String, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base


class Shift(Base):
    __tablename__ = "Shift"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    restaurantId: Mapped[str] = mapped_column(String, default="trump")
    username: Mapped[str] = mapped_column(String, nullable=False)
    role: Mapped[str] = mapped_column(String, default="waiter")
    status: Mapped[str] = mapped_column(String, default="active")
    startedAt: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    endedAt: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    startedBy: Mapped[str] = mapped_column(String, default="")
    endedBy: Mapped[str] = mapped_column(String, default="")
    endReason: Mapped[str] = mapped_column(String, default="")
    assignedTables: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    ordersHandled: Mapped[int] = mapped_column(Integer, default=0)
    revenueHandled: Mapped[float] = mapped_column(Float, default=0)
    responseMetrics: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    metadata_: Mapped[dict | None] = mapped_column("metadata", JSONB, nullable=True)
    createdAt: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updatedAt: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())


class AuditLog(Base):
    __tablename__ = "AuditLog"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    restaurantId: Mapped[str] = mapped_column(String, default="trump")
    actor: Mapped[str] = mapped_column(String, default="system")
    actorRole: Mapped[str] = mapped_column(String, default="")
    action: Mapped[str] = mapped_column(String, nullable=False)
    targetType: Mapped[str] = mapped_column(String, default="")
    targetId: Mapped[str] = mapped_column(String, default="")
    summary: Mapped[str] = mapped_column(String, default="")
    reason: Mapped[str] = mapped_column(String, default="")
    metadata_: Mapped[dict | None] = mapped_column("metadata", JSONB, nullable=True)
    createdAt: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


class Notification(Base):
    __tablename__ = "Notification"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    restaurantId: Mapped[str] = mapped_column(String, default="trump")
    source: Mapped[str] = mapped_column(String, nullable=False)
    title: Mapped[str] = mapped_column(String, nullable=False)
    body: Mapped[str] = mapped_column(String, default="")
    priority: Mapped[int] = mapped_column(Integer, default=3)
    recipientRole: Mapped[str] = mapped_column(String, default="")
    recipientUser: Mapped[str] = mapped_column(String, default="")
    tableId: Mapped[str] = mapped_column(String, default="")
    readAt: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    metadata_: Mapped[dict | None] = mapped_column("metadata", JSONB, nullable=True)
    createdAt: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


class WaiterTask(Base):
    __tablename__ = "WaiterTask"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    restaurantId: Mapped[str] = mapped_column(String, default="trump")
    tableId: Mapped[str] = mapped_column(String, default="")
    waiterName: Mapped[str] = mapped_column(String, default="")
    type: Mapped[str] = mapped_column(String, nullable=False)
    title: Mapped[str] = mapped_column(String, nullable=False)
    message: Mapped[str] = mapped_column(String, default="")
    priority: Mapped[int] = mapped_column(Integer, default=3)
    status: Mapped[str] = mapped_column(String, default="open")
    payload: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    requestedBy: Mapped[str] = mapped_column(String, default="system")
    approvedBy: Mapped[str] = mapped_column(String, default="")
    dueAt: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    acknowledgedAt: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    resolvedAt: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    createdAt: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updatedAt: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
